orca-retrofit-reactive/src/main/kotlin/com/netflix/spinnaker/orca/reactive/WindowedRequestMetrics.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2018 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.reactive
+
+import com.google.common.base.Ticker
+import com.google.common.cache.CacheBuilder
+import org.springframework.http.HttpStatus
+import java.time.Duration
+import java.util.concurrent.TimeUnit.MILLISECONDS
+
+/**
+ * A mechanism for recording request metrics that has a look-back limit. That is
+ * the metrics it reports represent those reported over the last [expireAfter]
+ * duration.
+ */
+class WindowedRequestMetrics
+@JvmOverloads constructor(
+  override val name: String,
+  private val expireAfter: Duration,
+  private val ticker: Ticker = Ticker.systemTicker()
+) : WritableRequestMetrics {
+
+  private object Unused
+
+  private val cache = CacheBuilder.newBuilder()
+    .expireAfterWrite(expireAfter.toMillis(), MILLISECONDS)
+    .ticker(ticker)
+    .weakKeys() // stops identical values colliding, we want all entries
+    .build<RequestMetric, Unused>()
+
+  // I'm sure this is really efficient
+  override val averageDuration: Duration
+    get() = cache.asMap().keys.map { it.duration }.average()
+
+  override val errorPercentage: Double
+    get() = when (cache.size()) {
+      0L   -> 0.0
+      else -> cache.run {
+        // We have to create a new collection as `Cache.size` is only weakly
+        // consistent and can report a different value to the actual number of
+        // unexpired keys. This is also true of the map returned by `toMap`.
+        ArrayList(asMap().keys).percentMatching { it.status.isError }
+      }
```

Comment:
error percentage is determined by the HTTP status of requests. 400+ is considered an error.

Reply:
Fair. Maybe just 429?

Reply:
429 also isn't an error and is actually an expected/common result with the way that the `RequestQueue` stuff works in clouddriver.

Return 429 so that a request is attempted elsewhere ... after some max period of time than it'll just failboat.

The failboat would be the thing to consider badness.